libutils/glob_lib.c

```diff
@@ -0,0 +1,516 @@
+{
+    assert(path != NULL);
+
+    char b_copy[strlen(path) + 1];
+    strcpy(b_copy, path);
+    const char *const b_name = basename(b_copy);
+    return StringEqual(b_name, ".") || StringEqual(b_name, "..");
+}
+
+/**
+ * Strategically walk through filesystem looking for files matching pattern.
+ */
+static void WalkPattern(
+    const char *const root,      // Currently matched subpath
+    const Seq *const components, // Glob components
+    const size_t index,          // Index of next (unmatched) glob component
+    Seq *const matches)          // Currently found matches
+{
+    assert(root != NULL);
+    assert(components != NULL);
+    assert(matches != NULL);
+
+    const size_t num_components = SeqLength(components);
+    assert(num_components > 0);
+    if (index >= num_components)
+    {
+        /* We have found a match, because we have recursively matched every
+         * glob component to a file or directory.
+         */
+        char *const match = NormalizePath(root);
+        SeqAppend(matches, match);
+        return;
+    }
+
+    const char *const component = SeqAt(components, index);
+    if (StringEqual(component, "."))
+    {
+        // We already know that this directory exists.
+        WalkPattern(root, components, index + 1, matches);
+    }
+
+    char *pattern =
+        (StringEndsWith(root, FILE_SEPARATOR_STR))
+            ? StringConcatenate(2, root, component)
+            : StringConcatenate(3, root, FILE_SEPARATOR_STR, component);
+
+    Seq *const files = ListDir(root, NULL);
+    const size_t num_files = (files == NULL) ? 0 : SeqLength(files);
+
+    for (size_t i = 0; i < num_files; i++)
+    {
+        const char *filename = SeqAt(files, i);
+        if (BasenameIsParentOrCurrent(filename))
+        {
+            // Iterating '.' & '..' directories would cause infinite recursion.
+        }
+        else if (GlobMatch(pattern, filename))
+        {
+            WalkPattern(filename, components, index + 1, matches);
+        }
+    }
+
+    SeqDestroy(files);
+    free(pattern);
+}
+
+Seq *GlobFind(const char *const pattern)
+{
+    assert(pattern != NULL);
+
+    Seq *const matches = SeqNew(1, free);
+
+    struct stat sb;
+    if (stat(pattern, &sb) == 0)
+    {
+        // The pattern itself is a match. There is no need to walk it.
+        char *const match = SafeStringDuplicate(pattern);
+        SeqAppend(matches, match);
+        return matches;
+    }
+
+    Seq *const components = StringSplit(pattern, PATH_DELIMITER);
+    if (SeqLength(components) == 0)
+    {
+        return matches;
+    }
+
+    const char *const component = SeqAt(components, 0);
+    if (StringEqual(component, ""))
+    {
+        // This is an absolute path.
+        WalkPattern(FILE_SEPARATOR_STR, components, 1, matches);
+    }
+    else
+    {
+        // This is a relative path.
+        WalkPattern(".", components, 0, matches);
+    }
+
+    SeqDestroy(components);
+    SeqSort(matches, StrCmpWrapper, NULL);
```

Comment:
Not sure this will always provide the desired results. I'd skip the sorting or did it on a directory level.

Reply:
> By default, the pathnames are sorted in ascending collation order
- glob(3)

Not sure what this means exactly. However, not sorting them might break our acceptance tests, that are dependent on the order. But I can always to without sorting and see if it works.